test/recipes/15-test_out_option.t

```diff
@@ -0,0 +1,61 @@
+#! /usr/bin/env perl
+# Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the OpenSSL license (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+
+use strict;
+use warnings;
+
+use File::Spec;
+use OpenSSL::Test qw/:DEFAULT srctop_file/;
+use OpenSSL::Test::Utils;
+
+setup("test_out_option");
+
+plan skip_all => "'-out' option tests are not available on Windows or VMS"
+    if $^O =~ /^(VMS|MSWin32)$/;
+
+plan tests => 11;
+
+# The following patterns should be tested:
+#
+# path        dirname
+# /usr/       /
+# /           /
+# .           .
+# ..          .
+
+test_illegal_path('/usr/');
+test_illegal_path('/');
+test_illegal_path('.');
+test_illegal_path('..');
+
+# Test for trying to create a file in a non-exist directory
+my @chars = ("A".."Z", "a".."z", "0".."9");
+my $rand_path;
+$rand_path .= $chars[rand @chars] for 1..32;
+$rand_path .= "/test.pem";
+
+test_illegal_path($rand_path);
+test_legal_path('test.pem');
+unlink 'test.pem';
+
+sub test_illegal_path {
+    my ($path) = @_;
+
+    my $start = time();
+    is(run(app([ 'openssl', 'genrsa', '-out', $path, '16384'])), 0, "invalid output path: $path");
+    my $end = time();
+    # The above process should exit in 2 seconds if the path is not valid
+    ok($end - $start < 2, "check time consumed");
+}
+
+sub test_legal_path {
+    my ($path) = @_;
+
+    ok(run(app([ 'openssl', 'genrsa', '-out', $path, '2048'])), "valid output path: $path");
```

Comment:
I just found that this command doesn't run the checks you're looking for, for one simple reason, [this in `apps/genrsa.c`](https://github.com/openssl/openssl/blob/master/apps/genrsa.c#L47):

``` C
    {"out", OPT_OUT, 's', "Output the key to specified file"},
```

The `'s'` needs to be changed to `'>'` first...

Reply:
I identified the files where `OUT_OPT` has the wrong opt character, here's the diff to fix it:

``` diff
diff --git a/apps/genrsa.c b/apps/genrsa.c
index f147852902..aa3195707a 100644
--- a/apps/genrsa.c
+++ b/apps/genrsa.c
@@ -43,7 +43,7 @@ const OPTIONS genrsa_options[] = {
     {"3", OPT_3, '-', "Use 3 for the E value"},
     {"F4", OPT_F4, '-', "Use F4 (0x10001) for the E value"},
     {"f4", OPT_F4, '-', "Use F4 (0x10001) for the E value"},
-    {"out", OPT_OUT, 's', "Output the key to specified file"},
+    {"out", OPT_OUT, '>', "Output the key to specified file"},
     OPT_R_OPTIONS,
     {"passout", OPT_PASSOUT, 's', "Output file pass phrase source"},
     {"", OPT_CIPHER, '-', "Encrypt the output with any supported cipher"},
diff --git a/apps/sess_id.c b/apps/sess_id.c
index f0c76b1917..dde62b0bd4 100644
--- a/apps/sess_id.c
+++ b/apps/sess_id.c
@@ -29,7 +29,7 @@ const OPTIONS sess_id_options[] = {
     {"outform", OPT_OUTFORM, 'f',
      "Output format - default PEM (PEM, DER or NSS)"},
     {"in", OPT_IN, 's', "Input file - default stdin"},
-    {"out", OPT_OUT, 's', "Output file - default stdout"},
+    {"out", OPT_OUT, '>', "Output file - default stdout"},
     {"text", OPT_TEXT, '-', "Print ssl session id details"},
     {"cert", OPT_CERT, '-', "Output certificate "},
     {"noout", OPT_NOOUT, '-', "Don't output the encoded session info"},
```

Reply:
Ahh, I didn't notice the 'genrsa' has already checked the output file before generating the key. Going to fix this.